test/e2e/network/service.go

```diff
@@ -2517,7 +2518,7 @@ var _ = common.SIGDescribe("Services", func() {
 		}
 	})
 
-	ginkgo.It("should respect internalTrafficPolicy=Local Pod to Pod [Feature:ServiceInternalTrafficPolicy]", func() {
+	framework.It("should respect internalTrafficPolicy=Local Pod to Pod", framework.WithFeatureGate(features.ServiceInternalTrafficPolicy), func() {
```

Comment:
We could also accept "ServiceInternalTrafficPolicy" as string here. But then we loose the compile-time verification that the string is actually a valid feature gate. It would still get checked at runtime whenever the suite gets invokved.

The advantages would be:
- no need for the extra package import
- shorter
- no need to change anything here when a feature graduates and (later) gets removed from k8s.io/kubernetes/pkg/features (but see the KEP discussion how the test should get annotated for a stable feature)

Reply:
I prefer to maintain the link to the code, to avoid drifting

Reply:
That's also my preference. Using the pre-defined string constant also has the advantage that tools like gopls can reliably find tests that reference it.

Reply:
I extended the PR with sets of valid features and environments. The framework itself doesn't register any of those, this will have to be added elsewhere (test/e2e/environments and test/e2e/features?).

Reply:
I added pre-defined features and node features in their own shared packages (`test/e2e/feature`, `test/e2e/nodefeature`). The names were chosen so that using these constants reads well. Tests are not allowed to add their own features or node features (registry gets frozen after those pre-defined ones).

There is nothing for environments yet because that's currently unused.

We have to allow arbitrary labels - there are simply too many of those and they might be useful. Those arbitrary labels don't get registered. That leads to a loophole: `framework.WithLabel("Feature:SomeCustomFeature"` is currently valid, but should be `feature.SomeCustomFeature` with a corresponding entry in the shared package. I think we can check those custom labels and complain when one of the other functions should be used instead.

Reply:
We also need a way to verify that tests are registered properly before we can merge this. There are some existing verify scripts that need to be updated. My plan is to do the verification as much as possible inside the E2E suite itself:
- Validate labels when functions are called.
- Use https://github.com/onsi/ginkgo/issues/1225 once it is available to add `--verify-tests` which then can be called by a verify script.

Reply:
I've completely revamped how issues discovered during test registration are handled. Each E2E suite now records such problems, then at the end does a sanity check that there were no issues. This works for `-dry-run` and `-focus=some tests`, so it'll be impossible to miss.

A verify script also checks this as part of `pull-kubernetes-verify`.